### PR TITLE
Add warning on long Auto Type sequences

### PR DESCRIPTION
**Pull request by marcbone**
Auto Type now shows a warning when you try to  repeat something too often.
Also you can now repeat your password and username

I'm new to all this open source, transifex,qt and c++ stuff. So dont spare with criticism but be nice.
I didnt know how to get the message-strings away from the code and had some problems with the tests, so it would be better if somebody looks over it. Sorry for that.

## Description

Large delays and large repetitions arent ignored anymore, instead you get a Messagebox asking you if you want to continue or abort.

## Motivation and context

It fixes Bug #216 

## How has this been tested?

I changed autoType command to repeat it 101 times and executed the autotype command.

Problem is, if you have multiple commands which are repeated very often (ex. {Password 101} {Username 101}) you get multiple MessageBoxes. I could have changed that, but I didnt find a way to do so without bad code or bigger changes.

Tested on Ubuntu 16.04

## Screenshots (if appropriate):
bildschirmfoto vom 2017-05-08 16-28-48
## Types of changes

- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document.
- ✅ My code follows the code style of this project. 
-  All new and existing tests passed. GUI-Test failed, bui it also failed without my changes so I guess its ok
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. (I have testesd coverage and directly the autotype features, diddnt get warnings on the console, but I really dont know what I should do there exactly

### CODE REVIEW

**Comment by TheZ3ro** [comment]
No criticism, every pull request is always appreciated.

Anyway like I said in this comment https://github.com/keepassxreboot/keepassxc/pull/217#issuecomment-275346013 this type of error reporting is broken-by-design. 
Autotype needs to be fast because users shouldn't wait and if users changes the window focus the autotype will fail.
This behavior blocks the execution of the autotype sequence and change the focus.

I will prefer adding a fairly simple syntax checker that checks when the users is writing the sequence.
Maybe I will develop this solution for 2.3.0 or a later 2.2.x

**Comment by droidmonkey** [comment]
I actually like this implementation a lot. But should also add a prompt on creation to warn about the downsides to excessive autotyping.

File: src/autotype/AutoType.cpp [hunk 1]
```diff
@@ -326,7 +326,9 @@ bool AutoType::parseActions(const QString& sequence, const Entry* entry, QList<A
                 return false;
             }
             else if (ch == '}') {
-                actions.append(createActionFromTemplate(tmpl, entry));
+                QList<AutoTypeAction*> autoType = createActionFromTemplate(tmpl, entry);
+                if (autoType.isEmpty()) return false;
```
Comment (TheZ3ro) [inline src/autotype/AutoType.cpp]: Use correct code-style

**Comment by TheZ3ro** [comment]
This actually "works" but every time a user execute the autotype it will ask Yes or No.

**Comment by marcbone** [comment]
Personally I dont think this is a problem if you have to insert your over 100 tabs or repeat your pw that often there is probabbly something wrong with what youre doing.

I'm more concerended of the problem mentioned in #217 about, switching windows and insert passwords were you dont want to insert them. So I would do an error message instead when trying to execute (maybe also in the creation).

Also I noticed the multiple pw, username insert gets ignored in keepass2.

File: src/autotype/AutoType.cpp [hunk 2]
```diff
@@ -383,11 +385,24 @@ QList<AutoTypeAction*> AutoType::createActionFromTemplate(const QString& tmpl, c
         // some safety checks
         else if (tmplName.compare("delay",Qt::CaseInsensitive)==0) {
             if (num > 10000) {
-                return list;
+                QMessageBox::StandardButton reply;
+                reply = QMessageBox::question(0,"AutoType",
```
Comment (phoerious) [inline src/autotype/AutoType.cpp]: You need to wrap your strings into tr() calls to make them translatable and please use spaces after commas.

File: src/autotype/AutoType.cpp [hunk 3]
```diff
@@ -383,11 +385,24 @@ QList<AutoTypeAction*> AutoType::createActionFromTemplate(const QString& tmpl, c
...
+                                              "This AutoType command contains a very long delay. Do you really want to execute it?");
+
+                if (reply==QMessageBox::No) {
```
Comment (phoerious) [inline src/autotype/AutoType.cpp]: Use spaces between operators

**Comment by droidmonkey** [comment]
My preference is to merge this in when the "on creation" check is added. Recommend defer to 2.3.0

**Comment by marcbone** [comment]
On creation check is added(error on wrong syntax, warnings for high delays and much repetiotions). Multiple warning problem on execution is fixed. 
The syntax checker is checking als functions which are [available in keepass2](http://keepass.info/help/base/autotype.html) also if they arent implemented in keepassxc so you dont get compatibility issues (I hope).

Only thing whats left is that the detection of too much repetiotion only works if you checked high delays first because it just checks if there is a whitespace followed by at least 3 digits.

File: src/gui/entry/EditEntryWidget.cpp [hunk 4]
```diff
@@ -478,11 +479,41 @@ void EditEntryWidget::updateEntryData(Entry *entry) const
...
+
+        QRegExp autoTypeSyntax
+            ("("+shortcutKeys + "|" + fixedStrings + "|\\{(" + normalCommands + "|" + specialLiterals + "|" + functionKeys
+                 + "|" + numpad + "|" + delay + "|" + beep + "|" + vkey+")\\})*");
         autoTypeSyntax.setCaseSensitivity(Qt::CaseInsensitive);
         autoTypeSyntax.setPatternSyntax(QRegExp::RegExp);
 
 
+
+        //small test @TODO delete
```
Comment (droidmonkey) [inline src/gui/entry/EditEntryWidget.cpp]: Ummmm

**Comment by marcbone** [comment]
autotype test is running because I switched the dialogs to databaseWidget. But it will only appear if you perform the autotype from the context menu. if you do it with the global autotype command you are landing in AutoTypeSelectDialog.cpp, but I don't know if I can get it to show the dialogs there, because the functions are using emits, something I've never heard of.

**Comment by droidmonkey** [comment]
emit is a Qt keyword

**Comment by TheZ3ro** [comment]
@marc Is this still under development or can I continue it myself?

**Comment by marcbone** [comment]
you can continue. I dont know how to fix the issue with the tests

**Comment by TheZ3ro** [comment]
@marcbone thanks for the contribution, I will cherry-pick your commits :)

**Comment by TheZ3ro** [comment]
Closing in favor of #1174

**Pull request closed without merging**